cmd/zrepl/zrepl.c

```diff
@@ -186,25 +187,24 @@ zio_cmd_free(zvol_io_cmd_t **cmd)
 
 
 static int
-uzfs_zvol_socket_read(int fd, char *buf, uint64_t nbytes)
+uzfs_zvol_socket_read(int fd, char *buf, uint64_t nbytes, int flags)
 {
 	uint64_t count = 0;
 	char *p = buf;
-	ZREPL_ERRLOG("Trying to read nbytes: %lu\n", nbytes);
 	while (nbytes) {
 		count = read(fd, (void *)p, nbytes);
-		if ((count <= 0) && (errno == EAGAIN)) {
+		if ((count <= 0) && (flags & O_NONBLOCK) &&
```

Comment:
good one

Reply:
I can't imagine a use case where this would be the right thing to do (O_NONBLOCK) in our code in a way it was designed. Could you explain in which situations we don't want to block on read or write?

Reply:
@jkryl ....uzfs_zvol_mgmt_thread connect to iSCSI controller and marking FD non-blocking. In future this thread may need to talk to multiple iSCSI controllers(on storage box can serve volume to many iSCSI controllers) so for each such connection socket will be created which are marked non--blocking. In such case, read/write on such FDs may come out without reading/writing data from/to it. So caller has take a call to pass flag if socket is marked non-blocking.

Reply:
But for the use case you have mentioned you would need to switch to event loop - Have one event loop with one thread which serves all iscsi targets, which is entirely different from the way how it is done now when for most of the things we use blocking calls + threads. In the current code O_NONBLOCK makes more harm than good. Consider what happens here. You cannot read data from socket so you try to read them again and again .. until available. That is a busy loop which does nothing useful but only drains CPU.